### PR TITLE
Add TFM support for stm32l562e_dk board

### PR DESCRIPTION
**Pull request by yestin**
This PR is to add the TF-M support for the stm32l562e_dk board.

### CODE REVIEW

File: boards/arm/stm32l562e_dk/stm32l562e_dk_ns.dts
Comment (yestin) [inline boards/arm/stm32l562e_dk/stm32l562e_dk_ns.dts]: This PR works but is not ideal. After TZ is enabled, the log output over usart1 is garbage from the non secure Zephyr application. I have used lpuart1 as a workaround but it is a bit awkward because I have to jump wire to wire out the lpuart1 TX/RX pins on stm32l562e_dk board, and the logs for mcuboot and secure firmware come out of usart1 while the log for non secure application comes out of lpuart1.

I have tried to avoid the GPIOA pin control in non secure target since GPIOA is a Trustzone protected peripheral after TZEN is enabled.

I ran the TFM sample application from the STM32Cube_FW_L5 package, the logging seems to work in non secure image, but I don't see anything special in the UART(COM) port initialization code.

After I switch to lpuart1, the logging in non secure application is working, but I don't understand the different since both lpuart1 and usart1 are securable peripherals, and are non-secure after reset.

I have spent quite some time to debug this issue but no avail so far. Maybe can @erwango @ABOSTM shed some light on this?

File: samples/tfm_integration/tfm_ipc/boards/stm32l562e_dk_ns.overlay
Comment (erwango) [inline samples/tfm_integration/tfm_ipc/boards/stm32l562e_dk_ns.overlay:11]: After some search it seems that this is not the case anymore:
https://github.com/zephyrproject-rtos/zephyr/pull/30210

It might be worth checking potential impacts on partition sizes

File: boards/arm/stm32l562e_dk/stm32l562e_dk_ns.dts
Comment (erwango) [reply boards/arm/stm32l562e_dk/stm32l562e_dk_ns.dts]: @yestin from what I see this is specific to this dk board and this works on the nucleo. I guess this is due to console confugration for this board on TFM side. Maybe @jamike you can confirm this ?

Comment (erwango) [reply boards/arm/stm32l562e_dk/stm32l562e_dk_ns.dts]: @yestin just to be sure: have you tried to keep uart1 as NS console **without pinmuxing enabled**

Comment (yestin) [reply boards/arm/stm32l562e_dk/stm32l562e_dk_ns.dts]: I believe this commit ([84d97dd8](https://github.com/zephyrproject-rtos/zephyr/pull/31118/commits/84d97dd894571bb0993c437e8effd71c69242f38)) shall disable the pinmuxing for the uart1 as NS console. After build, I checked the generated dts file build/zephyr/zephyr.dts and did not see the pinmux configuration.

File: samples/tfm_integration/tfm_ipc/boards/stm32l562e_dk_ns.overlay
Comment (yestin) [reply samples/tfm_integration/tfm_ipc/boards/stm32l562e_dk_ns.overlay:11]: I see. And yes, the non secure image in this sample is pretty small, like 19KB. I can try to decrease the non secure image size, but may still not enough for the secondary partitions.

File: boards/arm/stm32l562e_dk/stm32l562e_dk_ns.dts
Comment (erwango) [reply boards/arm/stm32l562e_dk/stm32l562e_dk_ns.dts]: > I believe this commit ([84d97dd8](https://github.com/zephyrproject-rtos/zephyr/pull/31118/commits/84d97dd894571bb0993c437e8effd71c69242f38)) shall disable the pinmuxing for the uart1 as NS console. After build, I checked the generated dts file build/zephyr/zephyr.dts and did not see the pinmux configuration.

And does it allow to get working console on uart1 ?

Comment (erwango) [reply boards/arm/stm32l562e_dk/stm32l562e_dk_ns.dts]: ok, I had a test and I confirm I'm seeing the same thing.

Comment (erwango) [reply boards/arm/stm32l562e_dk/stm32l562e_dk_ns.dts]: Though, at first sight, I'd rather suspect a clock issue.

Comment (erwango) [reply boards/arm/stm32l562e_dk/stm32l562e_dk_ns.dts]: So I manage to make ns console work on uart1 by setting `current-speed = <57600>;`.
This confirms it is a clock issue:
TFM and Zephyr doesn't configure the same clock tree for usart1.
Since we don't face this issue on lpuart1, I suspect the misalignment either comes from APB2 PRESC or USART1SEL(RCC_BDCR).

Comment (yestin) [reply boards/arm/stm32l562e_dk/stm32l562e_dk_ns.dts]: This is good stuff. I was able to see the ns console working with your suggestion.

Unfortunately, I think I lost my stm32l562e_dk board when I was trying to deactivate TZEN. I was able to recover it last time by applying 3V3 on PH3-BOOT0 signal to make it boot from RSS, but not this time. 😞 

I guess setting RDP level to 1 before TZEN deactivation was not a good idea. Anyway, I will not be able to debug any further on this for some time.

Comment (erwango) [reply boards/arm/stm32l562e_dk/stm32l562e_dk_ns.dts]: I faced some issue with the board myself during the investigation.
I managed to get it back using stm32cubeprogrammer, by erasing the flash fully and then programming a hello world (using stm32cubeprogrammer UI).

Comment (erwango) [reply boards/arm/stm32l562e_dk/stm32l562e_dk_ns.dts]: Should be fixed by https://github.com/zephyrproject-rtos/zephyr/pull/31178

File: samples/tfm_integration/tfm_ipc/boards/stm32l562e_dk_ns.overlay
Comment (microbuilder) [reply samples/tfm_integration/tfm_ipc/boards/stm32l562e_dk_ns.overlay:11]: I was just looking at this PR, and indeed with 512 KB flash it will be a bit of a juggling-act to try to make this work with the dual-bank bootloader. It's possible, but more complex projects will be a challenge without using a different approach for the secure bootloader.

File: boards/arm/stm32l562e_dk/stm32l562e_dk_ns.dts
Comment (yestin) [reply boards/arm/stm32l562e_dk/stm32l562e_dk_ns.dts]: Thanks for the fix. I have rebased on top of your fix and verified the new patch that the logging works well on usart1 now.

File: samples/tfm_integration/tfm_ipc/boards/stm32l562e_dk_ns.overlay
Comment (yestin) [reply samples/tfm_integration/tfm_ipc/boards/stm32l562e_dk_ns.overlay]: In the new patch, I leave the partition layout as is except updated the comments. As the secure image is almost 224 KB, it could not fit a secondary partition in 512 KB flash.

**Review by ioannisg — APPROVED** [review]
porting looks fine to me, and along the lines of the other tf-m capable platforms.
needs formal approval from @erwango

**Review by erwango — APPROVED** [review]

**Comment by galak** [comment]
@yestin do you mind rebasing this to address the merge conflict.

**Comment by ioannisg** [comment]
@yestin may i kindly ask you to do one more rebase of this PR? I recently merged #32044 which centralizes the signing for TFM builds, so you do not need to duplicate this any more in the board definition. Please, take a look at the other TFM-capable boards, for guidance.

After this rebase, assuming I get a +1 from @microbuilder , I'll merge this PR.

**Comment by yestin** [comment]
> @yestin may i kindly ask you to do one more rebase of this PR? I recently merged #32044 which centralizes the signing for TFM builds, so you do not need to duplicate this any more in the board definition. Please, take a look at the other TFM-capable boards, for guidance.
> 
> After this rebase, assuming I get a +1 from @microbuilder , I'll merge this PR.

Yes, I will look into rebasing on top of the new changes. I am running into issue that the signing generates .hex file instead .bin file with the new merge. The TFM_UPDATE.sh script calls STM32_Programmer_CLI to flash, and when the download file is hex file, it does not honor the downloading address that passed in.

The .hex file for NS image still links against the start address at 0x08000000, which is not a correct start address in secure state. For stm32l552xx and stm32l562xx, the flash memory space is aliased to 0x0C000000 in secure state (TZEN=1). I still need some time to debug this.

**Comment by yestin** [comment]
@ioannisg @microbuilder the rebased changes are ready for review. I add an option to the signing script to adjust the hex base address. This can allow the output hex file to be flashed to a different address. In the case for nucleo_l552ze_q and stm32l562e_dk boards, they have a different flash memory space (0x0C000000) when TZ is enabled.

I also update the nucleo_l552ze_q board to leverage the centralized signing code. cc: @erwango

**Review by erwango — APPROVED** [review]

**Comment by yestin** [comment]
@ioannisg do you get a chance to review the PR? It has been pending for some time and I am worried I would have to rebase it again. Please let me know if you have concerns.

**Comment by ioannisg** [comment]
> @ioannisg do you get a chance to review the PR? It has been pending for some time and I am worried I would have to rebase it again. Please let me know if you have concerns.

@yestin apologies for forgetting this PR, for some time. To clarify, the reason for adding DNM flag is that i was not able to run the tfm_ipc test properly, on the STM32L562 board, using the PR, so needed to investigate this more.

@yestin are you able to successfully compile, flash and run the tfm_ipc sample on this board _and_ on nucleo_l552ze_q?

File: boards/arm/nucleo_l552ze_q/doc/nucleol552ze_q.rst
Comment (erwango) [inline boards/arm/nucleo_l552ze_q/doc/nucleol552ze_q.rst:283]: Can you review the command here?
I had to go in `samples/tfm_integration/tfm_ipc/build` and then launch `west flash --hex-file tfm_merged.hex`

**Review by erwango — CHANGES_REQUESTED** [review]

**Review by erwango — APPROVED** [review]
Approved, I'll review the doc part once merged.

File: boards/arm/nucleo_l552ze_q/doc/nucleol552ze_q.rst
Comment (ioannisg) [reply boards/arm/nucleo_l552ze_q/doc/nucleol552ze_q.rst:283]: Yes, this needs an improvement.

File: boards/arm/stm32l562e_dk/stm32l562e_dk_ns_defconfig
Comment (ioannisg) [inline boards/arm/stm32l562e_dk/stm32l562e_dk_ns_defconfig:39]: We should probably enable HW_STACK_PROTECTION=y here as well. Or is there a problem with this @yestin @erwango ?

**Review by microbuilder — APPROVED** [review]

File: boards/arm/nucleo_l552ze_q/doc/nucleol552ze_q.rst
Comment (yestin) [reply boards/arm/nucleo_l552ze_q/doc/nucleol552ze_q.rst:283]: Looks like the build output is generated to the sample project directory instead of the root directory. I did a clean run and can see the merged hex file is generated at `build/tfm_merged.hex`. Do you have any local environments that may impact?

**Comment by yestin** [comment]
> > @ioannisg do you get a chance to review the PR? It has been pending for some time and I am worried I would have to rebase it again. Please let me know if you have concerns.
> 
> @yestin apologies for forgetting this PR, for some time. To clarify, the reason for adding DNM flag is that i was not able to run the tfm_ipc test properly, on the STM32L562 board, using the PR, so needed to investigate this more.
> 
> @yestin are you able to successfully compile, flash and run the tfm_ipc sample on this board _and_ on nucleo_l552ze_q?

Yes, I verify that I can run the tfm_ipc sample on both boards.

File: boards/arm/nucleo_l552ze_q/doc/nucleol552ze_q.rst
Comment (erwango) [reply boards/arm/nucleo_l552ze_q/doc/nucleol552ze_q.rst:283]: @yestin the file is present here indeed, though, when running the command, I'm getting the following west error:
```
$ west flash --hex-file build/stm32l562e_dk_ns/tfm_merged.hex
FATAL ERROR: --build-dir was not given, /local/mcu/zephyrproject/zephyr/samples/tfm_integration/tfm_ipc is not a build directory and the default build directory cannot be determined. Check your build.dir-fmt configuration option
```
@ioannisg is that what you're seeing as well ?

Comment (erwango) [reply boards/arm/nucleo_l552ze_q/doc/nucleol552ze_q.rst:283]: @yestin I confirm this is only due to my local build.dir-fmt configuration.
I've raised https://github.com/zephyrproject-rtos/zephyr/issues/33498 to discuss it

**Pull request merged**
